### PR TITLE
Ra 179 - Drag&Drop TreeViewBase and PlaylistTreeView

### PR DESCRIPTION
**Pull request by vmartos**

### CODE REVIEW

File: VAS.Core/Common/DragContextData.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program; if not, write to the Free Software
+//  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
+//
+using System;
+namespace VAS.Core.Common
+{
+	public struct DragContextData
```
Comment (ylatuya) [inline VAS.Core/Common/DragContextData.cs]: I would name it DragContext, since it's a context holding data in one of its properties instead.

File: VAS.Core/Common/RangeObservableCollection.cs [hunk 2]
```diff
@@ -38,17 +41,26 @@ public RangeObservableCollection (IEnumerable<T> collection) : base (collection)
...
+		/// <param name="items">Items to add</param>
 		public void AddRange (IEnumerable<T> items)
 		{
 			if (items.Any ()) {
+				int index = Items.Count;
 				foreach (T item in items) {
 					Items.Add (item);
 				}
-				NotifyCollectionChangedEventArgs e = new NotifyCollectionChangedEventArgs (NotifyCollectionChangedAction.Add, items.ToList ());
+				NotifyCollectionChangedEventArgs e = new NotifyCollectionChangedEventArgs (NotifyCollectionChangedAction.Add, items.ToList (), index);
```
Comment (ylatuya) [inline VAS.Core/Common/RangeObservableCollection.cs]: Can you add unit tests for it?

File: VAS.Core/Common/RangeObservableCollection.cs [hunk 3]
```diff
@@ -58,6 +70,31 @@ public void RemoveRange (IEnumerable<T> items)
 			OnCollectionChanged (e);
 		}
 
+		/// <summary>
+		/// Inserts multiple items at a specified position
+		/// </summary>
+		/// <param name="index">Position</param>
+		/// <param name="items">Items to add</param>
+		public void InsertRange (int index, IEnumerable<T> items)
```
Comment (ylatuya) [inline VAS.Core/Common/RangeObservableCollection.cs]: Can you add unit tests?

File: VAS.Core/Common/RangeObservableCollection.cs [hunk 4]
```diff
@@ -58,6 +70,31 @@ public void RemoveRange (IEnumerable<T> items)
...
+		{
+			try {
```
Comment (ylatuya) [inline VAS.Core/Common/RangeObservableCollection.cs]: This shouldn't be inside a try and catch

File: VAS.UI.Gtk2/UI/Common/TreeViewBase.cs [hunk 5]
```diff
@@ -79,16 +89,39 @@ public virtual void SetViewModel (object viewModel)
...
+		{
+			filter = new TreeModelFilter (store, null);
+			filter.VisibleFunc = new TreeModelFilterVisibleFunc (HandleFilter);
+			sort = new TreeModelSort (filter);
+			sort.SetSortFunc (COL_DATA, HandleSort);
+			sort.SetSortColumnId (COL_DATA, SortType.Ascending);
+			Model = sort;
+		}
+
+		protected void CreateDragSource (TargetEntry [] targetEntries)
```
Comment (ylatuya) [inline VAS.UI.Gtk2/UI/Common/TreeViewBase.cs]: Drag source is not need since qe create the drag begin our self

File: VAS.Services/Controller/PlaylistController.cs [hunk 6]
```diff
@@ -80,11 +80,11 @@ public void Start ()
 			App.Current.EventsBroker.SubscribeAsync<DeletePlaylistEvent> (HandleDeletePlaylist);
 			App.Current.EventsBroker.Subscribe<RenderPlaylistEvent> (HandleRenderPlaylist);
 			App.Current.EventsBroker.Subscribe<LoadPlaylistElementEvent> (HandleLoadPlaylistElement);
-
 			App.Current.EventsBroker.Subscribe<OpenedProjectEvent> (HandleOpenedProjectChanged);
 			App.Current.EventsBroker.Subscribe<LoadEventEvent> (HandleLoadPlayEvent);
 			App.Current.EventsBroker.Subscribe<LoadCameraEvent> (HandleLoadCameraEvent);
 			App.Current.EventsBroker.Subscribe<TimeNodeChangedEvent> (HandlePlayChanged);
+			App.Current.EventsBroker.Subscribe<UpdateEvent<Playlist>> (HandleUpdatePlaylist);
```
Comment (ylatuya) [inline VAS.Services/Controller/PlaylistController.cs]: Is this related to DnD?

File: VAS.Services/ViewModel/PlaylistCollectionVM.cs [hunk 7]
```diff
@@ -30,6 +34,45 @@ public class PlaylistCollectionVM : CollectionViewModel<Playlist, PlaylistVM>
...
+			});
+		}
+
+		/// <summary>
+		/// Moves the playlist elements, from different PlaylistVM to a unique destination PlaylistVM
+		/// </summary>
+		/// <param name="elementsToRemove">Elements to remove.</param>
+		/// <param name="elementsToAdd">Elements to add.</param>
+		/// <param name="index">Index.</param>
+		public void MovePlaylistElements (Dictionary<PlaylistVM, IEnumerable<PlaylistElementVM>> elementsToRemove,
```
Comment (ylatuya) [inline VAS.Services/ViewModel/PlaylistCollectionVM.cs]: The playlist controller handle this logic

File: VAS.UI.Gtk2/UI/Common/TreeViewBase.cs [hunk 8]
```diff
@@ -79,16 +89,39 @@ public virtual void SetViewModel (object viewModel)
...
+		{
+			enableDragSource = true;
+			targets = new TargetList (targetEntries);
+			EnableModelDragSource (ModifierType.None, targetEntries, DragAction.Default);
```
Comment (ylatuya) [inline VAS.UI.Gtk2/UI/Common/TreeViewBase.cs]: We don't need to enable DragSource in the model since we start the drags ourself with Drag Begin

File: VAS.Core/Common/RangeObservableCollection.cs (hunk 4 above)
Comment (vmartos) [inline VAS.Core/Common/RangeObservableCollection.cs]: Yep, I didn't remember to remove it.

File: VAS.Services/Controller/PlaylistController.cs (hunk 6 above)
Comment (vmartos) [inline VAS.Services/Controller/PlaylistController.cs]: Yes! because the ViewModel was the responsible of editing the playlist and i need an event to save the playlist in the controller. But as the next comment suggest, I can handle the Playlist editing logic in the controller and then this event won't be needed anymore.

File: VAS.UI.Gtk2/UI/Common/TreeViewBase.cs (hunk 8 above)
Comment (vmartos) [inline VAS.UI.Gtk2/UI/Common/TreeViewBase.cs]: If I don't do EnableModelDragSource I always have a Gtk.Critical when trying to Gtk.Drag.Begin

File: VAS.Services/ViewModel/PlaylistCollectionVM.cs [hunk 9]
```diff
@@ -40,6 +49,30 @@ public void LoadPlaylist (PlaylistVM playlist, PlaylistElementVM elementToStart,
...
+		{
+			int indexCopy = index;
+			for (int i = 0; i < indexCopy; i++) {
+				if (elementsToAdd.Value.Contains (elementsToAdd.Key.ViewModels [i])) {
+					index--;
+				}
+			}
+			foreach (var playlist in elementsToRemove) {
+				playlist.Key.ViewModels.RemoveRange (playlist.Value);
+				App.Current.EventsBroker.Publish (new UpdateEvent<Playlist> { Object = playlist.Key.Model, Force = true });
```
Comment (ylatuya) [inline VAS.Services/ViewModel/PlaylistCollectionVM.cs]: You could convert this into a Save() command in PlaylistVM

Comment (vmartos) [inline VAS.Services/ViewModel/PlaylistCollectionVM.cs]: This is outdated, now the PlaylisCollectionVM sends an event to move the elements and the PlaylistController handle this event to move them and save them.

File: VAS.UI.Gtk2/UI/Common/TreeViewBase.cs [hunk 10]
```diff
@@ -237,24 +271,92 @@ protected override bool OnButtonPressEvent (Gdk.EventButton evnt)
...
+			App.Current.DragContext.SourceData = draggedViewModels;
+		}
+
+		protected override void OnDragDataReceived (DragContext context, int x, int y, SelectionData selectionData, uint info, uint time)
+		{
+			List<IViewModel> draggedElements = App.Current.DragContext.SourceData as List<IViewModel>;
+			bool success = HandleDragReceived (draggedElements, x, y, Gtk.Drag.GetSourceWidget (context) == this);
+			Gtk.Drag.Finish (context, success, false, time);
+			App.Current.DragContext.SourceDataType = null;
+			App.Current.DragContext.SourceData = null;
```
Comment (ylatuya) [inline VAS.UI.Gtk2/UI/Common/TreeViewBase.cs]: You should probably reset this in drag end too, becuase this will only be triggered id the drag is successful

File: VAS.UI.Gtk2/UI/Common/TreeViewBase.cs (hunk 8 above)
Comment (ylatuya) [inline VAS.UI.Gtk2/UI/Common/TreeViewBase.cs]: ok

File: VAS.UI.Gtk2/UI/Common/TreeViewBase.cs (hunk 10 above)
Comment (vmartos) [inline VAS.UI.Gtk2/UI/Common/TreeViewBase.cs]: As i Remember the OnDragEnd not always is triggered, but I'll check that.
